### PR TITLE
[WIP] Couchdb support for everything

### PR DESCRIPTION
**Pull request by tduehr**
Add support for couchDb as a data backend for AUP, consent, gauth, etc..

### CODE REVIEW

File: api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java [hunk 1]
```diff
@@ -75,7 +84,8 @@ public CasEvent() {
      *
      * @return the creation time
      */
-    public ZonedDateTime getCreationTime() {
+    @JsonIgnore
+    public ZonedDateTime getZonedCreationTime() {
```
Comment (mmoayyed) [inline api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java]: `getCreationZonedDateTime`.

File: api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java [hunk 2]
```diff
@@ -37,27 +40,33 @@
...
 
+    @JsonProperty("creationTime")
     @Column(nullable = false)
     private String creationTime;
 
+    @JsonProperty("properties")
     @ElementCollection
     @MapKeyColumn(name = "name")
     @Column(name = "value")
-    @CollectionTable(name = "events_properties", joinColumns = @JoinColumn(name = "id"))
```
Comment (mmoayyed) [inline api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java]: Why this change?

File: api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java [hunk 3]
```diff
@@ -129,22 +139,27 @@ public void putAgent(final String dev) {
...
     }
 
+    @JsonIgnore
     public String getAgent() {
         return get("agent");
     }
 
-    public String getId() {
-        return get("id");
+    @JsonIgnore
```
Comment (mmoayyed) [inline api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java]: Why ignore?

File: core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java [hunk 4]
```diff
@@ -448,6 +487,48 @@ public static MultiValueMap asMultiValueMap(final String key1, final Object valu
...
+     */
+    @Deprecated
+    public static <T, C extends Collection<T>> C castCollection(final Class<T> klass, final Collection<? extends T> collection,
+                                                                            final Collector<T, ?, C> collector) {
+        return collection.stream().map(klass::cast).collect(collector);
+    }
+
+    /**
+     * Cast elements of a list.
+     * @deprecated temporary until interfaces can change
```
Comment (mmoayyed) [inline core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java]: Remove all deprecated functionality please.

File: core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java [hunk 5]
```diff
@@ -97,6 +98,44 @@
         return c;
     }
 
+    /**
+     * Convert the object given into a {@link List} instead.
+     *
+     * @param obj the object to convert into a collection
+     * @return The collection instance containing the object provided
+     */
+    public static List<Object> toList(final Object obj) {
```
Comment (mmoayyed) [inline core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java]: This seems to be duplicating the toCollection method.

File: settings.gradle [hunk 6]
```diff
@@ -149,6 +151,7 @@ include "support:cas-server-support-scim"
 include "support:cas-server-support-gua"
 include "support:cas-server-support-surrogate-api"
 include "support:cas-server-support-surrogate-authentication"
+include "support:cas-server-support-surrogate-authentication-couchb"
```
Comment (mmoayyed) [inline settings.gradle]: `couchdb`

File: gradle/dependencies.gradle [hunk 7]
```diff
@@ -1881,6 +1881,12 @@ ext.libraries = [
                     force = true
                 }
         ],
+
+        pac4jCouchDb             : dependencies.create("org.pac4j:pac4j-couch:$pac4jVersion") {
```
Comment (mmoayyed) [inline gradle/dependencies.gradle]: Please include this into the pac4j dependency.

File: support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/config/CasAcceptableUsagePolicyCouchDbConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    @Qualifier("defaultTicketRegistrySupport")
+    private TicketRegistrySupport ticketRegistrySupport;
+
+    @Autowired
+    private CasConfigurationProperties casProperties;
+
+    @Autowired
+    private ObjectMapperFactory objectMapperFactory;
+
+    @Bean
```
Comment (mmoayyed) [inline support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/config/CasAcceptableUsagePolicyCouchDbConfiguration.java]: Please include `ConditionalOnMissingBean` annotations for all beans.

Comment (mmoayyed) [reply support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/config/CasAcceptableUsagePolicyCouchDbConfiguration.java]: Also consider using ObjectProvider for injected fields.

File: support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/consent/CouchDbConsentRepository.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * This is {@link CouchDbConsentRepository}.
+ *
+ * @author Timur Duehr
+ * @since 6.0.0
+ */
+@Getter
+@Setter
+@AllArgsConstructor
```
Comment (mmoayyed) [inline support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/consent/CouchDbConsentRepository.java]: RequiredArgsConstructor would be better, if you make fields be final.

File: support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbConsentDecision.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        setAttributes(c.getAttributes());
+        setPrincipal(c.getPrincipal());
+        setCreatedDate(c.getCreatedDate());
+        setId(c.getId());
+        setOptions(c.getOptions());
+        setReminder(c.getReminder());
+        setReminderTimeUnit(c.getReminderTimeUnit());
+        setService(c.getService());
+        if (getId() < 0) {
+            setId(SecureRandom.getInstanceStrong().nextLong());
```
Comment (mmoayyed) [inline support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbConsentDecision.java]: nextLong could return a negative, right?

Comment (mmoayyed) [reply support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbConsentDecision.java]: You may want to avoid using SecureRandom natively, and use the CAS wrapper class instead.

File: support/cas-server-support-couchdb-authentication/src/main/java/org/apereo/cas/authentication/CouchDbAuthenticationHandler.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.Getter;
+import lombok.Setter;
+
+/**
+ * This is {@link CouchDbAuthenticationHandler}.
+ *
+ * @author Timur Duehr
+ * @since 6.0.0
+ */
+@Getter
```
Comment (mmoayyed) [inline support/cas-server-support-couchdb-authentication/src/main/java/org/apereo/cas/authentication/CouchDbAuthenticationHandler.java]: Annotations should be removed here, if they are no-ops?

File: support/cas-server-support-couchdb-authentication/src/main/java/org/apereo/cas/config/CouchDbAuthenticationConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+    @ConditionalOnMissingBean(name = "couchDbPrincipalFactory")
+    @Bean
+    public PrincipalFactory couchDbPrincipalFactory() {
+        return PrincipalFactoryUtils.newPrincipalFactory();
+    }
+
+    @Bean
+    @RefreshScope
+    public AuthenticationHandler couchDbAuthenticationHandler() {
```
Comment (mmoayyed) [inline support/cas-server-support-couchdb-authentication/src/main/java/org/apereo/cas/config/CouchDbAuthenticationConfiguration.java]: `@ConditionalOnMissingBean`?

File: build.gradle [hunk 13]
```diff
@@ -90,6 +90,14 @@ allprojects {
     apply plugin: "signing"
     apply plugin: "jacoco"
     apply plugin: "groovy"
+    apply plugin: "pmd"
+
+    pmd {
```
[the remark this answers is not in the document]
Comment (tduehr) [reply build.gradle]: I actually didn't mean for that to land in this PR at all...

File: api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java [hunk 14]
```diff
@@ -66,6 +69,15 @@ public CasEvent() {
         this.id = System.currentTimeMillis();
     }
 
+
+    public String getCreationTimeRaw() {
```
[the remark this answers is not in the document]
Comment (tduehr) [reply api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java]: Was to prevent a collision in the JSON. has since been resolved in another manner...

File: api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java [hunk 15]
```diff
@@ -66,6 +69,15 @@ public CasEvent() {
...
+        return creationTime;
+    }
+
+    public long getIdRaw() {
```
[the remark this answers is not in the document]
Comment (tduehr) [reply api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java]: As above...

File: support/cas-server-support-audit-couchdb/src/main/java/org/apereo/cas/audit/CouchDbAuditTrailManager.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        if (asyncronous) {
+            this.executorService.execute(() -> couchDb.add(new CouchDbAuditActionContext(auditActionContext)));
+        } else {
+            couchDb.add(new CouchDbAuditActionContext(auditActionContext));
+        }
+    }
+
+    @Override
+    public Set<AuditActionContext> getAuditRecordsSince(final LocalDate localDate) {
+        return Sets.newHashSet(couchDb.findAuditRecordsSince(localDate));
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-audit-couchdb/src/main/java/org/apereo/cas/audit/CouchDbAuditTrailManager.java]: Done

File: support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/aup/CouchDbAcceptableUsagePolicyRepository.java [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+ * This is {@link CouchDbAcceptableUsagePolicyRepository}.
+ *
+ * @author Timur Duehr
+ * @since 5.3.0
+ */
+@Slf4j
+public class CouchDbAcceptableUsagePolicyRepository extends AbstractPrincipalAttributeAcceptableUsagePolicyRepository {
+
+    private static final long serialVersionUID = -2391630070546362552L;
+    private ProfileCouchDbRepository couchDb;
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/aup/CouchDbAcceptableUsagePolicyRepository.java]: done

File: support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/aup/CouchDbAcceptableUsagePolicyRepository.java [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+            for (int retries = 0; retries < conflictRetries; retries++) {
+                try {
+                    exception = null;
+                    couchDb.update(profile);
+                    success = true;
+                } catch (final UpdateConflictException e) {
+                    exception = e;
+                }
+                if (success) {
+                    LOGGER.debug("Successfully deleted service [{}].", profile.getUsername());
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/aup/CouchDbAcceptableUsagePolicyRepository.java]: fixed

File: support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/config/CasConsentCouchDbConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+
+/**
+ * This is {@link CasConsentCouchDbConfiguration}.
+ *
+ * @author Timur Duehr
+ * @since 5.3.0
+ */
+public class CasConsentCouchDbConfiguration {
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/config/CasConsentCouchDbConfiguration.java]: Done

File: support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/consent/CouchDbConsentRepository.java [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * This is {@link CouchDbConsentRepository}.
+ *
+ * @author Timur Duehr
+ * @since 5.3.0
+ */
+@Getter
+@Setter
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/consent/CouchDbConsentRepository.java]: Done

File: support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbConsentDecision.java [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+
+    /**
+     * Copy consent details to this instance.
+     * @param other decision to copy details from.
+     * @return CouchDb capable consent decision.
+     */
+    @SneakyThrows
+    public CouchDbConsentDecision copyDetailsFrom(final ConsentDecision other) {
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbConsentDecision.java]: We need to retain `cid` and `rev`. I'll see if there's a better way...

File: support/cas-server-support-events-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbCasEvent.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        this.cid = cid;
+        this.rev = rev;
+    }
+
+    /**
+     * Copy constructor.
+     * @param event Event to copy from.
+     */
+    @SneakyThrows
+    public CouchDbCasEvent(final CasEvent event) {
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-events-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbCasEvent.java]: just calls super now.

File: support/cas-server-support-events-couchdb/src/main/java/org/apereo/cas/support/events/CouchDbCasEventRepository.java [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+/**
+ * This is {@link CouchDbCasEventRepository}.
+ *
+ * @author Timur Duehr
+ * @since 5.3.0
+ */
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-events-couchdb/src/main/java/org/apereo/cas/support/events/CouchDbCasEventRepository.java]: Done

File: support/cas-server-support-gauth-couchdb/src/main/java/org/apereo/cas/adaptors/gauth/GoogleAuthenticatorCouchDbTokenRepository.java [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+import java.time.LocalDateTime;
+
+/**
+ * This is {@link GoogleAuthenticatorCouchDbTokenRepository}.
+ *
+ * @author Timur Duehr
+ * @since 5.3.0
+ */
+@Slf4j
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-gauth-couchdb/src/main/java/org/apereo/cas/adaptors/gauth/GoogleAuthenticatorCouchDbTokenRepository.java]: Done

File: support/cas-server-support-gauth-couchdb/src/main/java/org/apereo/cas/adaptors/gauth/GoogleAuthenticatorCouchDbTokenRepository.java [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Override
+    public OneTimeToken get(final String uid, final Integer otp) {
+        return couchDb.findOneByUidForOtp(uid, otp);
+    }
+
+    @Override
+    protected void cleanInternal() {
+        try {
+            couchDb.findByIssuedDateTimeBefore(LocalDateTime.now().minusSeconds(expireTokensInSeconds)).forEach(couchDb::remove);
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-gauth-couchdb/src/main/java/org/apereo/cas/adaptors/gauth/GoogleAuthenticatorCouchDbTokenRepository.java]: Done

File: support/cas-server-support-gauth-couchdb/src/main/java/org/apereo/cas/config/GoogleAuthenticatorCouchDbConfiguration.java [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.context.config.annotation.RefreshScope;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.annotation.EnableScheduling;
+
+/**
+ * This is {@link GoogleAuthenticatorCouchDbConfiguration}.
+ *
+ * @author Misagh Moayyed
```
[the remark this answers is not in the document]
Comment (tduehr) [reply support/cas-server-support-gauth-couchdb/src/main/java/org/apereo/cas/config/GoogleAuthenticatorCouchDbConfiguration.java]: Done

File: api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java (hunk 1 above)
Comment (tduehr) [reply api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java]: Done.

File: api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java (hunk 3 above)
Comment (tduehr) [reply api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java]: These are calculated fields based on the contents of other fields that are going into the JSON. They are not required to persist the object.

File: api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java (hunk 2 above)
Comment (tduehr) [reply api/cas-server-core-api-events/src/main/java/org/apereo/cas/support/events/dao/CasEvent.java]: because I changed `getId()` to avoid collision with the `id` field.

File: core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java (hunk 5 above)
Comment (tduehr) [reply core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/CollectionUtils.java]: very similar... it's needed to rebuild a `Set` into a `List` and cast the elements. I will rework the interface in another PR to get rid of this. Currently, only `public List<String> getEligibleAccountsForSurrogateToProxy(final String username)` uses this function. A Set will do just fine but requires redoing the interface.

File: gradle/dependencies.gradle (hunk 7 above)
Comment (tduehr) [reply gradle/dependencies.gradle]: Why pull in every pac4j data service when we only need one? I only added the one dependency spec because the rest aren't relevant to this.

File: settings.gradle (hunk 6 above)
Comment (tduehr) [reply settings.gradle]: Done... That one took me a while to find yesterday...

File: support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/config/CasAcceptableUsagePolicyCouchDbConfiguration.java (hunk 8 above)
Comment (tduehr) [reply support/cas-server-support-aup-couchdb/src/main/java/org/apereo/cas/config/CasAcceptableUsagePolicyCouchDbConfiguration.java]: Will do.

File: support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/consent/CouchDbConsentRepository.java (hunk 9 above)
Comment (tduehr) [reply support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/consent/CouchDbConsentRepository.java]: Done

File: support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbConsentDecision.java (hunk 10 above)
Comment (tduehr) [reply support/cas-server-support-consent-couchdb/src/main/java/org/apereo/cas/couchdb/CouchDbConsentDecision.java]: Done.

File: support/cas-server-support-couchdb-authentication/src/main/java/org/apereo/cas/authentication/CouchDbAuthenticationHandler.java (hunk 11 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-authentication/src/main/java/org/apereo/cas/authentication/CouchDbAuthenticationHandler.java]: Done

File: support/cas-server-support-couchdb-authentication/src/main/java/org/apereo/cas/config/CouchDbAuthenticationConfiguration.java (hunk 12 above)
Comment (tduehr) [reply support/cas-server-support-couchdb-authentication/src/main/java/org/apereo/cas/config/CouchDbAuthenticationConfiguration.java]: Done

**Comment by tduehr** [comment]
PR broken to individual components.
